[data] Handle missing 'chunks' key when Databricks UC query returns zero rows

Toshaksha opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Fixes a bug where querying a Databricks UC table that returns zero rows causes a KeyError due to the missing "chunks" key in the manifest.  
This change checks for the "chunks" key and returns an empty datasource instead of crashing, improving user experience for empty query results.

## Related issue number

Closes #54369 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

Toshaksha committed:
Handle missing 'chunks' key when Databricks UC query returns zero rows to avoid crash

Signed-off-by: Toshaksha <147024929+Toshaksha@users.noreply.github.com>

Toshaksha committed:
Prevent sending auth headers when fetching external URLs

Signed-off-by: Toshaksha <147024929+Toshaksha@users.noreply.github.com>

gvspraveen commented on python/ray/data/_internal/datasource/databricks_uc_datasource.py:
```diff
@@ -99,81 +99,95 @@ def __init__(
+            def empty_read_task(task_index, parallelism):
+                def empty_read_fn():
+                    if False:
+                        yield    # to make this a generator
```
is this better to yield a empty Pyarrow table here ?

```

# Yield empty PyArrow table for empty dataset
                    yield pyarrow.table({})
                    ```

gvspraveen commented:
@Toshaksha Looks reasonable. See my one suggestion above. Also do you want to update unit tests?  [Here is the existing test setup for this file](https://github.com/ray-project/ray/blob/master/python/ray/data/tests/test_sql.py#L149), which exercises similar logic.

Also there are lint failures in the build. Can you please address those?

Toshaksha committed:
Fix empty read task to yield empty PyArrow table

Signed-off-by: Toshaksha <147024929+Toshaksha@users.noreply.github.com>

Toshaksha committed:
Add test for Databricks UC with empty result set

Signed-off-by: Toshaksha <147024929+Toshaksha@users.noreply.github.com>

Toshaksha commented:
Hi @gvspraveen 👋,

Thanks again for your earlier suggestions!

- I updated the empty_read_task to yield an empty PyArrow table (pyarrow.table({})) as you recommended.
-  Also added a unit test covering the empty result scenario for Databricks UC.
-  All commits are signed off and formatting was handled via scripts/format.sh.

However, the Buildkite microcheck job failed due to an infrastructure issue, not the PR code itself. Here's the error:

DENIED: Adding tag '25d99372-forge' to an image in the repository with name 'rayproject/citemp' in registry with id '029272617770' exceeds the maximum allowed number of tags per image which is '1000'

Let me know if there's anything else I should tweak. Thanks!

Toshaksha

gvspraveen committed:
Merge branch 'master' into ray_chunks

gvspraveen commented:
@Toshaksha I also notice some lint failures. can u please address them?

https://buildkite.com/ray-project/microcheck/builds/20145#01980567-7823-4a56-a6df-aa2384c439d0/185-282

Toshaksha committed:
Merge branch 'master' into ray_chunks

gvspraveen commented on python/ray/data/_internal/datasource/databricks_uc_datasource.py:
```diff
@@ -99,81 +99,94 @@ def __init__(
+                return ReadTask(read_fn=read_fn, metadata=metadata)
 
-        self._get_read_task = get_read_task
+            self._get_read_task = get_read_task
```
You also need to fix `def get_read_tasks(self, parallelism: int)` method below to handle case where `self.num_chunks == 0`. Otherwise there wont be any read tasks created.

One thing you can do in that method is :

```
# Handle empty dataset case
        if self.num_chunks == 0:
            return [self._get_read_task(0, 1)]

        if parallelism > self.num_chunks:
        .....
```

Toshaksha committed:
Handle empty result set in DatabricksUCDatasource

Signed-off-by: Toshaksha <toshaksha02@gmail.com>

Toshaksha commented:
Hi @gvspraveen 👋,

All suggested changes have been made:
- Empty result now yields an empty pyarrow.table({})
- get_read_tasks() updated to handle num_chunks == 0
- Lint issues fixed via scripts/format.sh

One test job (data: arrow v20 3.12) exited with code 42, but all tests, including test_databricks_uc_datasource_empty_result, passed successfully. The logs show the exit happened during artifact cleanup (Directory not empty, cleaning up...), followed by a clean exit 0, so this looks infra-related and not caused by the PR.

That said, if there's anything else you'd like me to check or improve from my side, happy to follow up. Appreciate your review and guidance!

gvspraveen commented:
@Toshaksha Thanks for making these changes. It seems latest change doesnt have safety check while accessing dictionary.

This is causing [CI to fail on Keyerror](https://buildkite.com/ray-project/microcheck/builds/20366#01980f6c-9b51-4126-9420-a32dcb1943f9) with latest changes.

```
[2025-07-15T20:26:53Z]         if self.is_truncated:
--
  | [2025-07-15T20:26:53Z]             logger.warning(
  | [2025-07-15T20:26:53Z]                 f"The resulting size of the dataset of '{query!r}' exceeds "
  | [2025-07-15T20:26:53Z]                 "100GiB and it is truncated."
  | [2025-07-15T20:26:53Z]             )
  | [2025-07-15T20:26:53Z]
  | [2025-07-15T20:26:53Z] >       chunks = manifest["chunks"]
  | [2025-07-15T20:26:53Z] E       KeyError: 'chunks'
  | [2025-07-15T20:26:53Z]
  | [2025-07-15T20:26:53Z] /rayci/python/ray/data/_internal/datasource/databricks_uc_datasource.py:110: KeyError
```

Toshaksha committed:
Fix KeyError by using .get() for 'truncated' and 'chunks' in manifest

Signed-off-by: Toshaksha <147024929+Toshaksha@users.noreply.github.com>

Toshaksha commented:
Hi @gvspraveen, I’ve fixed the KeyError by using .get() for 'chunks' and 'truncated' in the manifest as you suggested. This should resolve the CI failure. Please let me know if there’s anything else!

gvspraveen commented:
@Toshaksha Unit test is still failing in CI. I took a stab and was able to get it to fix. You can see fixes in my [draft here](https://github.com/ray-project/ray/pull/55057). Which is on top of your changes.

Can you patch these changes on your PR and see if you can get CI to pass?

bveeramani committed:
Test commit

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

bveeramani committed:
Remove test commit

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

bveeramani committed:
Initial commit

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

bveeramani committed:
Remove comment

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

gvspraveen committed:
Fix CI and tests

Signed-off-by: Praveen Gorthy <praveeng@anyscale.com>

gvspraveen committed:
Merge branch 'master' into ray_chunks
